include/libtorrent/piece_picker.hpp

```diff
@@ -385,6 +385,10 @@ namespace libtorrent
 		// or if we have the piece
 		bool is_piece_finished(int index) const;
 
+		// returns true if all blocks in this piece are finished
+		// and have been written to disk
```

Comment:
did you mean to say "and the piece has passed the hash-check"?
if all blocks are ``finished``, that means the piece has been written to disk already, right? otherwise it would be in ``writing``, no?

Reply:
The finished state does *not* imply the piece has been verified. Hash checking and writing to disk are done asynchronously so to make this work there need to be checks in both the write complete and piece verified paths.

Reply:
Strike that last sentence, I missed that there is already a check in the verified path.

Reply:
I was also under the impression that the ``state_finished`` block state did not imply having passed the hash check. The ``downloading_piece`` has a ``passed_hash_check`` member, which I would expect would have to checked. But I haven't looked very deep

Reply:
I'm posting the alert right after piece_finished_alert, that alert implies that the piece has passed the hash check. And I *think* piece_picker::is_piece_finished() also implies that even though is not documented as such, but I'm not 100% sure. So I don't think I need to check passed_hash_check right now but I may have to after I implement your previous suggestion.

Reply:
I see. although, if a piece being "ready" means that it's written to disk and has passed the hash-check, it would be an error prone interface of the piece picker to say a piece is ready even though it hasn't passed the check. i.e. I think piece_picker should have a sensible interface, and not rely on specific external logic to behave well

Reply:
It's not. Right now calling is_piece_finished() guarantees that the piece has passed the hash check AFAIU. It would've been clearer if I had called has_piece_passed(). It will be much clearer after I implement your previous advice of removing the is_piece_finished() call as then I will probably have to check passed_hash_check too.

Reply:
Actually, you're absolutely right, I get it now. There is a change that is_piece_ready() will return a false positive. It just will never happen when I'm calling it. I fixed it already, will test later and push.